From 05d56bbe57eaf84b607bac7ebca090769f450ed1 Mon Sep 17 00:00:00 2001
From: Pratik Pachange <quic_ppachang@quicinc.com>
Date: Fri, 28 Mar 2025 09:00:03 +0000
Subject: [PATCH] v4l2: Send video alignment in allocation query

- Send video alignment in allocation query, so that upstream plugin can
  consume this information when allocation query is returned
- Send GstAllocationParams padding in allocation query

Upstream-Status: Inappropriate [Qualcomm specific]
Signed-off-by: Pratik Pachange <quic_ppachang@quicinc.com>
---
 sys/v4l2/gstv4l2object.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 4c4a334..39abc8e 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -5973,6 +5973,7 @@ gst_v4l2_object_propose_allocation (GstV4l2Object * obj, GstQuery * query)
   guint size, min, max;
   GstCaps *caps;
   gboolean need_pool;
+  GstAllocationParams a_params;
   GstStructure *allocation_meta = NULL;
 
   /* Set defaults allocation parameters */
@@ -6023,6 +6024,27 @@ gst_v4l2_object_propose_allocation (GstV4l2Object * obj, GstQuery * query)
 
   min = MAX (obj->min_buffers, GST_V4L2_MIN_BUFFERS (obj));
 
+  /* allocation params */
+  gst_allocation_params_init (&a_params);
+
+  if (V4L2_TYPE_IS_MULTIPLANAR (obj->type)) {
+    struct v4l2_pix_format_mplane *pix_mp = &obj->format.fmt.pix_mp;
+    gint i;
+
+    for (i = 0, size = 0; i < pix_mp->num_planes; i++)
+      size += pix_mp->plane_fmt[i].sizeimage;
+  } else {
+    size = obj->format.fmt.pix.sizeimage;
+  }
+
+  a_params.padding = size - GST_V4L2_SIZE (obj);
+
+  if (a_params.padding) {
+    GST_DEBUG_OBJECT (obj->dbg_obj, "allocation params padding:%" G_GSIZE_FORMAT,
+        a_params.padding);
+    gst_query_add_allocation_param (query, NULL, &a_params);
+  }
+
   gst_query_add_allocation_pool (query, pool, size, min, max);
 
   if (obj->align.padding_top || obj->align.padding_bottom ||
-- 
2.25.1

